platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/AggregatingProblemConsumer.java

```diff
@@ -0,0 +1,102 @@
+import org.gradle.internal.build.event.types.DefaultProblemAggregationDetails;
+import org.gradle.internal.build.event.types.DefaultProblemDescriptor;
+import org.gradle.internal.build.event.types.DefaultProblemEvent;
+import org.gradle.internal.operations.CurrentBuildOperationRef;
+import org.gradle.internal.operations.OperationIdentifier;
+import org.gradle.tooling.internal.protocol.InternalProblemAggregation;
+import org.gradle.tooling.internal.protocol.InternalProblemDetails;
+import org.gradle.tooling.internal.protocol.InternalProblemEvent;
+import org.gradle.tooling.internal.protocol.problem.InternalBasicProblemDetails;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+@NonNullApi
+public class AggregatingProblemConsumer {
+    private final Multimap<String, InternalProblemEvent> seenProblems = ArrayListMultimap.create();
+    private final ProgressEventConsumer progressEventConsumer;
+    private final Supplier<OperationIdentifier> operationIdentifierSupplier;
+
+    AggregatingProblemConsumer(ProgressEventConsumer progressEventConsumer, Supplier<OperationIdentifier> operationIdentifierSupplier) {
+        this.progressEventConsumer = progressEventConsumer;
+        this.operationIdentifierSupplier = operationIdentifierSupplier;
+    }
+
+    void sendProblemSummaries() {
+        List<InternalProblemAggregation> problemSummaries = createSummaries();
+
+        if (problemSummaries.isEmpty()) {
+            return;
+        }
+
+        DefaultProblemEvent event = new DefaultProblemEvent(
+            new DefaultProblemDescriptor(operationIdentifierSupplier.get(), CurrentBuildOperationRef.instance().getId()),
+            new DefaultProblemAggregationDetails(problemSummaries));
+
+        progressEventConsumer.progress(event);
+        seenProblems.clear();
+    }
+
+    private List<InternalProblemAggregation> createSummaries() {
+        return seenProblems.asMap().values()
+            .stream()
+            .map(ImmutableList::copyOf)
+            .filter(values -> values.size() > 1)
+            .map(aggregatedEvents -> {
+                InternalBasicProblemDetails firstProblem = (InternalBasicProblemDetails) aggregatedEvents.iterator().next().getDetails();
+                return new DefaultInternalProblemAggregation(firstProblem.getCategory(), firstProblem.getLabel(), aggregatedEvents);
+            })
+            .collect(toImmutableList());
+    }
+
+    void emit(InternalProblemEvent problem) {
+        InternalProblemDetails details = problem.getDetails();
+        if (!(details instanceof InternalBasicProblemDetails)) {
+            return;
+        }
+
+        InternalBasicProblemDetails basicDetails = (InternalBasicProblemDetails) details;
+        String aggregationKey = basicDetails.getCategory().getCategory() + ";" + basicDetails.getLabel().getLabel();
+
+        if(seenProblems.size() > 100_000) {
```

Comment:
❌ This is incorrect. The size() method will return the size of `keySet()` which means the collection can grow indefinitely as long as it gets values for the same keys. I think it's a simple-enough solution to maintain global counter of inserted entries and check on that.

Reply:
Actually not.
it returns the number of values according [the doc](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Multimap.html#size--).
I added tests to `AggregatingProblemConsumerTest.groovy` to verify that.